### PR TITLE
Migrate postgres user

### PR DESCRIPTION
**Pull request by tiffanyhan**
Resolve #4664

**AC:**
- [ ] Every time a User is created, they are also created in postgres
- [ ] Every time a User is updated, the update is propagated to postgres
- [ ]  a postdeploy hook is written to copy all Users to postgres

### CODE REVIEW

**Comment by tiffanyhan** [comment]
@mattkrick started over on a new branch for this, looking forward to your feedback!

File: packages/server/database/types/AuthToken.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import appOrigin from '../../appOrigin'
 interface Input {
   sub: string
   tms: string[]
-  rol?: 'su'
+  rol?: 'su' | null
```
Comment (tiffanyhan) [inline packages/server/database/types/AuthToken.ts:9]: todo

File: packages/server/graphql/mutations/helpers/bootstrapNewUser.ts [hunk 2]
```diff
@@ -41,10 +47,14 @@ const bootstrapNewUser = async (newUser: User, isOrganic: boolean) => {
...
-    user: r.table('User').insert(newUser),
-    event: r.table('TimelineEvent').insert(joinEvent)
-  }).run()
+
+  await Promise.all([
+    r({
+      user: r.table('User').insert(newUser),
+      event: r.table('TimelineEvent').insert(joinEvent)
+    }).run(),
+    catchAndLog(() => insertUserQuery.run((newUser as unknown) as IInsertUserQueryParams, getPg()))
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/helpers/bootstrapNewUser.ts:56]: tested

File: packages/server/graphql/intranetSchema/mutations/connectSocket.ts [hunk 3]
```diff
@@ -47,7 +50,11 @@ export default {
     }
     const datesAreOnSameDay = now.toDateString() === lastSeenAt?.toDateString()
     if (!datesAreOnSameDay) {
-      await db.write('User', userId, {inactive: false, lastSeenAt: now})
+      const dbUpdates = {inactive: false, lastSeenAt: now}
+      await Promise.all([
+        catchAndLog(() => updateUser(dbUpdates, userId)),
+        db.write('User', userId, dbUpdates)
+      ])
```
Comment (tiffanyhan) [inline packages/server/graphql/intranetSchema/mutations/connectSocket.ts:57]: tested

File: packages/server/postgres/queries/src/appendUserTmsQuery.sql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+/*
+  @name appendUserTmsQuery
+*/
+UPDATE "User" SET
+  tms = arr_append_uniq(tms, :teamId)
```
Comment (tiffanyhan) [inline packages/server/postgres/queries/src/appendUserTmsQuery.sql:5]: tested

File: packages/server/safeMutations/addTeamIdToTMS.ts [hunk 5]
```diff
@@ -1,19 +1,25 @@
...
-        .contains(teamId)
-        .default(false),
-      user('tms'),
-      user('tms')
-        .default([])
-        .append(teamId)
-    )
-  }))
+  return Promise.all([
+    catchAndLog(() => appendUserTmsQuery.run({id: userId, teamId}, getPg())),
```
Comment (tiffanyhan) [inline packages/server/safeMutations/addTeamIdToTMS.ts:10]: tested

File: packages/server/graphql/mutations/helpers/updateUserProfile.ts [hunk 6]
```diff
@@ -65,7 +67,8 @@ const updateUserProfile = async (
       .update(updates, {returnChanges: true})('changes')('new_val')
       .default([])
       .run() as unknown) as TeamMember[],
-    db.write('User', userId, updates)
+    db.write('User', userId, updates),
+    catchAndLog(() => updateUser(updates, userId))
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/helpers/updateUserProfile.ts:71]: tested

File: packages/server/graphql/mutations/helpers/removeTeamMember.ts [hunk 7]
```diff
@@ -122,7 +125,10 @@ const removeTeamMember = async (
     tms: user('tms').difference([teamId])
   })
 
-  const user = await db.write('User', userId, reqlUpdater)
+  const [user] = await Promise.all([
+    db.write('User', userId, reqlUpdater),
+    catchAndLog(() => removeUserTmsQuery.run({ids: [userId], teamIds: [teamId]}, getPg()))
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/helpers/removeTeamMember.ts:130]: tested

File: packages/server/billing/helpers/adjustUserCount.ts [hunk 8]
```diff
@@ -51,6 +53,7 @@ const changePause = (inactive: boolean) => async (_orgIds: string[], userId: str
     event: inactive ? 'Account Paused' : 'Account Unpaused'
   })
   return Promise.all([
+    catchAndLog(() => updateUser({inactive}, userId)),
```
Comment (tiffanyhan) [inline packages/server/billing/helpers/adjustUserCount.ts:56]: tested

File: packages/server/graphql/mutations/deleteUser.ts [hunk 9]
```diff
@@ -66,12 +68,23 @@ export default {
...
     })
     // do this after 30 seconds so any segment API calls can still get the email
+
     setTimeout(() => {
       db.write('User', userIdToDelete, {
         isRemoved: true,
         email: 'DELETED',
         reasonRemoved: validReason
       })
+      catchAndLog(() =>
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/deleteUser.ts:78]: tested

**Comment by tiffanyhan** [comment]
it looks like there are some valid sql statements which i've tested work in psql, but fails bc pgtyped can't parse them :/

Ex: `WHERE :id = true AND id IN (COALESCE(:idValue, '{}'));` works in psql, but fails with:

```SEND TO SENTRY TypeError: Cannot read property 'map' of null
    at Object.exports.processSQLQueryAST (/Users/tiffanyhan/resources/parabol/node_modules/@pgtyped/query/lib/preprocessor-sql.js:24:22)
    at PreparedQuery.<anonymous> (/Users/tiffanyhan/resources/parabol/node_modules/@pgtyped/query/lib/tag.js:36:76)
    at Generator.next (<anonymous>)```

Similarly, `WHERE :id = true AND id IN (CASE WHEN :idValue THEN :idValue ELSE '{}' END);` works in psql, but fails at the pg build generation step with:

```
Error in query. Details: {
  errorCode: 'boolin',
  hint: undefined,
  message: 'invalid input syntax for type boolean: "{}"',
  position: '855'
}
```

**Comment by tiffanyhan** [comment]
hey @mattkrick wondering if you found or thought of a better way to do static partial updates with support for setting values to null.

if not, i don't have much tension around writing a standalone query whenever we have to set a field to null. i think those cases are rarer and it'll simplify things greatly.

**Comment by mattkrick** [comment]
totally agree, single purpose write queries sound like a good plan!

File: packages/server/graphql/intranetSchema/mutations/connectSocket.ts [hunk 10]
```diff
@@ -47,7 +54,19 @@ export default {
...
-      await db.write('User', userId, {inactive: false, lastSeenAt: now})
+      await Promise.all([
+        catchAndLog(() =>
+          updateUserQuery.run(
+            ({
+              inactive: false,
+              lastSeenAt: now,
+              ids: [userId]
+            } as unknown) as IUpdateUserQueryParams,
+            getPg()
```
Comment (tiffanyhan) [inline packages/server/graphql/intranetSchema/mutations/connectSocket.ts:65]: tested

File: packages/server/graphql/mutations/helpers/updateUserProfile.ts [hunk 11]
```diff
@@ -65,7 +71,13 @@ const updateUserProfile = async (
       .update(updates, {returnChanges: true})('changes')('new_val')
       .default([])
       .run() as unknown) as TeamMember[],
-    db.write('User', userId, updates)
+    db.write('User', userId, updates),
+    catchAndLog(() =>
+      updateUserQuery.run(
+        (Object.assign(updates, {ids: [userId]}) as unknown) as IUpdateUserQueryParams,
+        getPg()
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/helpers/updateUserProfile.ts:78]: tested

File: packages/server/graphql/mutations/deleteUser.ts [hunk 12]
```diff
@@ -66,12 +72,24 @@ export default {
...
+        updateUserQuery.run(
+          ({
+            isRemoved: true,
+            email: `DELETED:${userId}:${new Date()}`,
+            reasonRemoved: validReason,
+            ids: [userIdToDelete]
+          } as unknown) as IUpdateUserQueryParams,
+          getPg()
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/deleteUser.ts:90]: tested

File: packages/server/billing/helpers/adjustUserCount.ts [hunk 13]
```diff
@@ -51,6 +57,15 @@ const changePause = (inactive: boolean) => async (_orgIds: string[], userId: str
     event: inactive ? 'Account Paused' : 'Account Unpaused'
   })
   return Promise.all([
+    catchAndLog(() =>
+      updateUserQuery.run(
+        ({
+          inactive,
+          ids: [userId]
+        } as unknown) as IUpdateUserQueryParams,
+        getPg()
```
Comment (tiffanyhan) [inline packages/server/billing/helpers/adjustUserCount.ts:66]: tested

File: packages/server/graphql/mutations/loginWithGoogle.ts [hunk 14]
```diff
@@ -75,7 +81,18 @@ const loginWithGoogle = {
...
-          await db.write('User', viewerId, {identities})
+          await Promise.all([
+            db.write('User', viewerId, {identities}),
+            catchAndLog(() =>
+              updateUserQuery.run(
+                ({
+                  identities,
+                  ids: [viewerId]
+                } as unknown) as IUpdateUserQueryParams,
+                getPg()
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/loginWithGoogle.ts:92]: tested

File: packages/server/graphql/intranetSchema/mutations/flagOverLimit.ts [hunk 15]
```diff
@@ -28,7 +35,20 @@ const flagOverLimit = {
...
+          updateUserQuery.run(
+            ({
+              overLimitCopy: copy,
+              ids: userIds
+            } as unknown) as IUpdateUserQueryParams,
+            getPg()
+          )
+      : () => clearOverLimitCopyQuery.run({ids: userIds}, getPg())
+    await Promise.all([
+      catchAndLog(pgQueryCb),
```
Comment (tiffanyhan) [inline packages/server/graphql/intranetSchema/mutations/flagOverLimit.ts:49]: tested

File: packages/server/graphql/mutations/emailPasswordReset.ts [hunk 16]
```diff
@@ -73,7 +79,19 @@ const emailPasswordReset = {
...
-      await db.write('User', userId, {identities})
+      const updates = {identities}
+      await Promise.all([
+        catchAndLog(() =>
+          updateUserQuery.run(
+            ({
+              updates,
+              ids: [userId]
+            } as unknown) as IUpdateUserQueryParams,
+            getPg()
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/emailPasswordReset.ts:90]: tested

File: packages/server/graphql/mutations/resetPassword.ts [hunk 17]
```diff
@@ -73,6 +79,15 @@ const resetPassword = {
       localIdentity.hashedPassword = await bcrypt.hash(newPassword, Security.SALT_ROUNDS)
       localIdentity.isEmailVerified = true
       await Promise.all([
+        catchAndLog(() =>
+          updateUserQuery.run(
+            ({
+              identities,
+              ids: [userId]
+            } as unknown) as IUpdateUserQueryParams,
+            getPg()
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/resetPassword.ts:88]: tested

File: packages/server/graphql/mutations/verifyEmail.ts [hunk 18]
```diff
@@ -60,7 +66,19 @@ export default {
...
-        await db.write('User', userId, {identities, updatedAt: now})
+        await Promise.all([
+          catchAndLog(() =>
+            updateUserQuery.run(
+              ({
+                identities,
+                updatedAt: now,
+                ids: [userId]
+              } as unknown) as IUpdateUserQueryParams,
+              getPg()
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/verifyEmail.ts:77]: tested

File: packages/server/safeMutations/addTeamIdToTMS.ts [hunk 19]
```diff
@@ -1,19 +1,25 @@
...
+    db.write('User', userId, (user) => ({
```
Comment (tiffanyhan) [inline packages/server/safeMutations/addTeamIdToTMS.ts:11]: tested

File: packages/server/graphql/mutations/helpers/removeTeamMember.ts (hunk 7 above)
Comment (tiffanyhan) [inline packages/server/graphql/mutations/helpers/removeTeamMember.ts:130]: tested

File: packages/server/safeMutations/safeArchiveTeam.ts [hunk 20]
```diff
@@ -10,9 +13,12 @@ const safeArchiveTeam = async (teamId: string) => {
...
     .filter({isNotRemoved: true})('userId')
     .run()
-  const users = await db.writeMany('User', userIds, (user) => ({
-    tms: user('tms').difference([teamId])
-  }))
+  const [users] = await Promise.all([
+    db.writeMany('User', userIds, (user) => ({
+      tms: user('tms').difference([teamId])
+    })),
+    catchAndLog(() => removeUserTmsQuery.run({ids: userIds, teamIds: [teamId]}, getPg()))
```
Comment (tiffanyhan) [inline packages/server/safeMutations/safeArchiveTeam.ts:20]: tested

File: packages/server/graphql/mutations/addFeatureFlag.ts [hunk 21]
```diff
@@ -58,10 +64,16 @@ export default {
...
-      featureFlags: user('featureFlags').default([]).append(flag).distinct()
+      featureFlags: user('featureFlags')
+        .default([])
+        .append(flag)
+        .distinct()
     })
     const userIds = users.map(({id}) => id)
-    await db.writeMany('User', userIds, reqlUpdater)
+    await Promise.all([
+      catchAndLog(() => appendUserFeatureFlagsQuery.run({ids: userIds, flag}, getPg())),
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/addFeatureFlag.ts:74]: tested

File: packages/server/graphql/intranetSchema/mutations/addNewFeature.ts [hunk 22]
```diff
@@ -43,7 +46,8 @@ const addNewFeature = {
         .table('NewFeature')
         .insert(newFeature)
         .run(),
-      db.writeTable('User', {newFeatureId})
+      db.writeTable('User', {newFeatureId}),
+      catchAndLog(() => addUserNewFeatureQuery.run({newFeatureId}, getPg()))
```
Comment (tiffanyhan) [inline packages/server/graphql/intranetSchema/mutations/addNewFeature.ts:50]: tested

File: packages/server/graphql/mutations/dismissNewFeature.ts [hunk 23]
```diff
@@ -10,7 +13,11 @@ export default {
   resolve: async (_source, _args, {authToken}) => {
     // AUTH
     const viewerId = getUserId(authToken)
-    await db.write('User', viewerId, {newFeatureId: null})
+    const update = {newFeatureId: null}
+    await Promise.all([
+      catchAndLog(() => dismissNewFeatureQuery.run({ids: [viewerId]}, getPg())),
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/dismissNewFeature.ts:18]: tested

File: packages/server/graphql/mutations/payLater.ts [hunk 24]
```diff
@@ -70,7 +73,10 @@ export default {
         .default(0)
         .add(1)
     })
-    await db.write('User', viewerId, reqlUpdater)
+    await Promise.all([
+      catchAndLog(() => incrementUserPayLaterClickCountQuery.run({id: viewerId}, getPg())),
```
Comment (tiffanyhan) [inline packages/server/graphql/mutations/payLater.ts:77]: tested

File: packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts [hunk 25]
```diff
@@ -0,0 +1,186 @@
+import {R} from 'rethinkdb-ts'
+import User from '../types/User'
+
```
Comment (tiffanyhan) [inline packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts:3]: copy and pasted from previous ran migration

**Comment by tiffanyhan** [comment]
@mattkrick ready for review! i merged branches so this PR address all AC from #4664

File: packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts [hunk 26]
```diff
@@ -106,15 +106,63 @@ export const up = async function(r: R) {
...
-      .group('email')
-      .count()
-      .ungroup()
-      .filter((row) => row('reduction').gt(1))('group')
-      .filter((row) => row.ne('DELETED'))
-      .run({arrayLimit: 200000})) as string[]
-    
+    // hard coded bc we don't want to risk running the agg query in prod
+    // may need to get a fresh list before running backfill in prod
+    const affectedEmails = [
```
Comment (mattkrick) [inline packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts:111]: we can't have any PII in our source code. can you do a hard reset on this commit?

consider storing this list in a DB table or a JSON file that is imported during CI

File: packages/server/graphql/intranetSchema/mutations/connectSocket.ts [hunk 27]
```diff
@@ -47,7 +54,19 @@ export default {
...
     if (!datesAreOnSameDay) {
-      await db.write('User', userId, {inactive: false, lastSeenAt: now})
+      await Promise.all([
+        catchAndLog(() =>
+          updateUserQuery.run(
+            ({
+              inactive: false,
+              lastSeenAt: now,
+              ids: [userId]
+            } as unknown) as IUpdateUserQueryParams,
```
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/graphql/intranetSchema/mutations/connectSocket.ts]: that is such a good point. i really like that! helper function added.

File: packages/server/graphql/intranetSchema/mutations/flagOverLimit.ts [hunk 28]
```diff
@@ -28,7 +35,20 @@ const flagOverLimit = {
...
+    const pgQueryCb = copy
+      ? () =>
+          updateUserQuery.run(
+            ({
+              overLimitCopy: copy,
+              ids: userIds
+            } as unknown) as IUpdateUserQueryParams,
+            getPg()
+          )
+      : () => clearOverLimitCopyQuery.run({ids: userIds}, getPg())
```
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/graphql/intranetSchema/mutations/flagOverLimit.ts]: ah true true! i added and double checked setting to empty string will work.

File: packages/server/graphql/mutations/helpers/bootstrapNewUser.ts [hunk 29]
```diff
@@ -41,10 +44,14 @@ const bootstrapNewUser = async (newUser: User, isOrganic: boolean) => {
...
-    user: r.table('User').insert(newUser),
-    event: r.table('TimelineEvent').insert(joinEvent)
-  }).run()
+
+  await Promise.all([
+    r({
+      user: r.table('User').insert(newUser),
+      event: r.table('TimelineEvent').insert(joinEvent)
+    }).run(),
+    catchAndLog(() => insertUserQuery.run(newUser as unknown as IInsertUserQueryParams, getPg()))
```
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/graphql/mutations/helpers/bootstrapNewUser.ts]: necessary bc it complains that type of `identities` which is `AuthIdentity[]` is different from type json. added a helper function for greater type safety

File: packages/server/postgres/migrations/postDeploy/1613857272859_backfillUserTable.ts [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+  const doBackfill = async (
+    usersByFieldChoice: ('createdAt' | 'updatedAt'),
+    usersAfterTs?: Date
+  ) => {
+    let i = 0
+    let foundUsers = false
+
+    console.log('starting backfill pass...')
+    console.log('after ts:', usersAfterTs, 'by:', usersByFieldChoice)
+    while (true) {
```
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/postgres/migrations/postDeploy/1613857272859_backfillUserTable.ts]: oh that's an interesting point, changed!

File: packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts [hunk 31]
```diff
@@ -0,0 +1,186 @@
...
+
+  try {
+    const affectedEmails = (await r
+      .table('User')
+      .group('email')
+      .count()
+      .ungroup()
+      .filter((row) => row('reduction').gt(1))('group')
+      .filter((row) => row.ne('DELETED'))
+      .run({arrayLimit: 200000})) as string[]
```
[the remark this answers is not in the document]
Comment (tiffanyhan) [reply packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts:116]: this refers to an outdated version of the file

File: packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts (hunk 26 above)
Comment (tiffanyhan) [reply packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts]: yes. force pushed a rebase. will get the emails from a DB table

**Comment by tiffanyhan** [comment]
@mattkrick changes have been made per your suggestions, thanks for giving this your attention

File: packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts [hunk 32]
```diff
@@ -0,0 +1,194 @@
...
+
+    // replace team invitations
+    await r
+      .table('TeamInvitation')
+      .filter({invitedBy: badUserId})
+      .update({invitedBy: goodUserId})
+      .run()
+    await r
+      .table('Notification')
+      .filter({
```
Comment (mattkrick) [inline packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts:90]: Notification has a userId index that should be used here

File: packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts [hunk 33]
```diff
@@ -0,0 +1,194 @@
...
+            },
+            {returnChanges: true}
+          )
+      })
+      .run()
+
+    await r
+      .table('User')
+      .getAll(r.args(toDeleteIds))
+      .delete({returnChanges: true})
```
Comment (mattkrick) [inline packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts:185]: no need for returnChanges?

File: packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts [hunk 34]
```diff
@@ -0,0 +1,194 @@
...
+          .table('User')
+          .get(update('id'))
+          .update(
+            {
+              email: update('email'),
+              picture: update('picture'),
+              tms: update('tms'),
+              identities: update('identities')
+            },
+            {returnChanges: true}
```
Comment (mattkrick) [inline packages/server/database/migrations/20210224121516-mergeDuplicateUsers.ts:177]: no need for returnChanges?

**Review by mattkrick — APPROVED** [review]
2 little returnChanges that we could remove & after that it's great! I'll let you do the honors of merging!

**Comment by tiffanyhan** [comment]
Closing in favor of #4787

**Pull request closed without merging**